spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/inbound/HazelcastMessageProducer.java

```diff
@@ -0,0 +1,106 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.hazelcast.inbound;
+
+import java.util.Set;
+
+import com.hazelcast.core.DistributedObject;
+
+import reactor.util.StringUtils;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.integration.endpoint.MessageProducerSupport;
+import org.springframework.integration.hazelcast.common.CacheListeningPolicyType;
+import org.springframework.integration.hazelcast.common.HazelcastIntegrationDefinitionValidator;
+import org.springframework.integration.hazelcast.common.HazelcastIntegrationUtils;
+import org.springframework.integration.hazelcast.context.ApplicationContextStartEventHandler;
+import org.springframework.messaging.Message;
+
+/**
+ * This is a base class for Hazelcast Event-Driven Message Producers.
+ * 
+ * @author Eren Avsarogullari
+ * @since 1.0.0
+ *
+ */
+public class HazelcastMessageProducer extends MessageProducerSupport implements BeanFactoryPostProcessor, DisposableBean {
+
+	private DistributedObject distributedObject;
+
+	private String cacheEventTypes;
+
+	private CacheListeningPolicyType cacheListeningPolicy;
+
+	@Override
+	protected void onInit() {
+		if (HazelcastIntegrationDefinitionValidator
+				.validateCacheEventByDistributedObject(getDistributedObject(), getCacheEventTypeSet())) {
+			super.onInit();
+		}
+	}
+
+	public Set<String> getCacheEventTypeSet() {
+		return StringUtils.commaDelimitedListToSet(getCacheEventTypes());
+	}
+
+	@Override
+	public void sendMessage(Message<?> message) {
+		super.sendMessage(message);
+	}
+
+	@Override
+	public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory) throws BeansException {
+		if (!beanFactory.containsBean(ApplicationContextStartEventHandler.class.getName())) {
+			beanFactory.registerSingleton(
+					ApplicationContextStartEventHandler.class.getName(),
+					new ApplicationContextStartEventHandler());
+		}
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		HazelcastIntegrationUtils.shutdownAllHazelcastInstances();
+	}
+
+	public DistributedObject getDistributedObject() {
+		return distributedObject;
+	}
+
+	public void setDistributedObject(DistributedObject distributedObject) {
+		this.distributedObject = distributedObject;
+	}
+
+	public String getCacheEventTypes() {
+		return cacheEventTypes;
+	}
+
+	public void setCacheEventTypes(String cacheEventTypes) {
+		this.cacheEventTypes = cacheEventTypes;
+	}
+
+	public CacheListeningPolicyType getCacheListeningPolicy() {
```

Comment:
When you rework the components dependency order, all these `public` getters should be `protected`